src/stpipe/library.py

```diff
@@ -0,0 +1,850 @@
+        """
+        names = set()
+        for member in self._members:
+            names.add(member["group_id"])
+        return names
+
+    @property
+    def group_indices(self):
+        """
+        A ``dict`` with "group_id" as keys. Each value contains
+        a ``list`` of indices of members that shared that "group_id".
+
+        Note that this is based on the member entries in the association
+        used to create the ModelLibrary. Updating the "group_id" of a model
+        in the library will NOT change the values returned by this property
+        (and will not change the association data available at ``asn``).
+        """
+        group_dict = {}
+        for i, member in enumerate(self._members):
+            group_id = member["group_id"]
+            if group_id not in group_dict:
+                group_dict[group_id] = []
+            group_dict[group_id].append(i)
+        return group_dict
+
+    def __len__(self):
+        return len(self._members)
+
+    def borrow(self, index):
+        """
+        "borrow" a model from the library.
+
+        Parameters
+        ----------
+        index : int
+            The index of the model within the library. As the library
+            size does not change this is unique to this model for this
+            library.
+
+        Returns
+        -------
+        model : DataModel
+
+        Raises
+        ------
+        `ClosedLibraryError`
+            If the library is not "open" (used in a ``with`` context).
+
+        `BorrowError`
+            If the model at this index is already "borrowed".
+        """
+        if not self._open:
+            raise ClosedLibraryError("ModelLibrary is not open")
+
+        # if model was already borrowed, raise
+        if index in self._ledger:
+            raise BorrowError("Attempt to double-borrow model")
+
+        # if this model is in memory, return it
+        if self._on_disk:
+            if index in self._temp_filenames:
+                model = self._datamodels_open(
+                    self._temp_filenames[index], **self._datamodels_open_kwargs
+                )
+            else:
+                model = self._load_member(index)
+        else:
+            if index in self._loaded_models:
+                model = self._loaded_models[index]
+            else:
+                model = self._load_member(index)
+                self._loaded_models[index] = model
+
+        self._ledger[index] = model
+        return model
+
+    def _temp_path_for_model(self, model, index):
+        """
+        Determine the temporary path to use to save a model
+        at the provided index in the library.
+        """
+        model_filename = self._model_to_filename(model)
+        subpath = self._temp_path / f"{index}"
+        if not os.path.exists(subpath):
+            os.makedirs(subpath)
+        return subpath / model_filename
+
+    def shelve(self, model, index=None, modify=True):
```

Comment:
when you borrow and then later shelve an "on-disk" model, does it get removed from memory?  I assumed it would, but I don't see (probably just missed it) where this removal happens

Reply:
The library retains no references to the model after its been shelved (and saved to the temporary file). As long as the step code also contains no references it will be available for garbage collection next time python runs a collection.